transcoder not returning errors to client properly

Sample log: 
2026/02/10 21:49:03 [INFO] StreamVideo: Transcoding required for /media/New Folder/video.flv
2026/02/10 21:49:03 [INFO] Starting background transcode: /media/New Folder/video.flv -> /cache/transcoded/video.flv_w0.mp4
2026/02/10 21:49:03 [INFO] Waiting for transcode to complete: /media/New Folder/video.flv
2026/02/10 21:49:03 [ERROR] Background transcode failed for /media/New Folder/video.flv: ffmpeg error: exit status 187 - ffmpeg version 8.0.1 Copyright (c) 2000-2025 the FFmpeg developers
  built with gcc 15.2.0 (Alpine 15.2.0)
  configuration: --prefix=/usr --disable-librtmp --disable-lzma --disable-static --disable-stripping --enable-avfilter --enable-gpl --enable-ladspa --enable-libaom --enable-libass --enable-libbluray --enable-libdav1d --enable-libdrm --enable-libfontconfig --enable-libfreetype --enable-libfribidi --enable-libharfbuzz --enable-libjxl --enable-libmp3lame --enable-libopenmpt --enable-libopus --enable-libplacebo --enable-libpulse --enable-librav1e --enable-librist --enable-libshaderc --enable-libsoxr --enable-libsrt --enable-libssh --enable-libtheora --enable-libv4l2 --enable-libvidstab --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx264 --enable-libx265 --enable-libxcb --enable-libxml2 --enable-libxvid --enable-libzimg --enable-libzmq --enable-lto=auto --enable-lv2 --enable-openssl --enable-pic --enable-pthreads --enable-shared --enable-vaapi --enable-vdpau --enable-version3 --enable-vulkan --optflags=-O3 --enable-libsvtav1 --enable-libvpl
  libavutil      60.  8.100 / 60.  8.100
  libavcodec     62. 11.100 / 62. 11.100
  libavformat    62.  3.100 / 62.  3.100
  libavdevice    62.  1.100 / 62.  1.100
  libavfilter    11.  4.100 / 11.  4.100
  libswscale      9.  1.100 /  9.  1.100
  libswresample   6.  1.100 /  6.  1.100
Input #0, flv, from '/media/New Folder/video.flv':
  Metadata:
    lasttimestamp   : 1386
    datasize        : 13037
    metadatacreator : Manitu Group FLV MetaData Injector 2
    canSeekToEnd    : false
    videosize       : 73276742
    audiosize       : 6392104
    lastkeyframetimestamp: 1385
    audiodelay      : 0
    creationdate    : unknown
    metadatadate    : 2010-07-15T12:57:28.878000Z
  Duration: 00:23:05.82, start: 0.000000, bitrate: 459 kb/s
  Stream #0:0: Audio: mp3 (mp3float), 22050 Hz, mono, fltp, 32 kb/s
  Stream #0:1: Video: vp6f, yuv420p, 648x459, 419 kb/s, 29.90 fps, 29.97 tbr, 1k tbn
Stream mapping:
  Stream #0:1 -> #0:0 (vp6f (native) -> h264 (libx264))
  Stream #0:0 -> #0:1 (mp3 (mp3float) -> aac (native))
Press [q] to stop, [?] for help
[libx264 @ 0x7f180d406040] height not divisible by 2 (648x459)
[vost#0:0/libx264 @ 0x7f180a7e1c80] [enc:libx264 @ 0x7f180e174a40] Error while opening encoder - maybe incorrect parameters such as bit_rate, rate, width or height.
[vf#0:0 @ 0x7f1808c61ac0] Error sending frames to consumers: Generic error in an external library
[vf#0:0 @ 0x7f1808c61ac0] Task finished with error code: -542398533 (Generic error in an external library)
[vf#0:0 @ 0x7f1808c61ac0] Terminating thread with return code -542398533 (Generic error in an external library)
[vost#0:0/libx264 @ 0x7f180a7e1c80] [enc:libx264 @ 0x7f180e174a40] Could not open encoder before EOF
[vost#0:0/libx264 @ 0x7f180a7e1c80] Task finished with error code: -22 (Invalid argument)
[vost#0:0/libx264 @ 0x7f180a7e1c80] Terminating thread with return code -22 (Invalid argument)
[out#0/mp4 @ 0x7f180f10a7c0] Nothing was written into output file, because at least one of its streams received no packets.
frame=    0 fps=0.0 q=0.0 Lsize=       0KiB time=N/A bitrate=N/A speed=N/A elapsed=0:00:00.11
[aac @ 0x7f180c265840] Qavg: 57610.211
Conversion failed!

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,6 +7,14 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 # Changelog
 
+## [0.12.2] - Unreleased
+
+### Fixed
+
+- **Video Dimension Compatibility**: Fixed transcoding failures for videos with unusual dimensions. Some older videos (particularly FLV files and certain codecs) have dimensions that aren't compatible with modern web video formats, causing "height not divisible by 2" errors and preventing playback. The transcoder now automatically adjusts video dimensions during conversion to ensure compatibility, allowing these videos to play successfully in your browser. ([#244](https://github.com/djryanj/media-viewer/issues/244))
+
+- **Transcoding Error Display**: Fixed transcoding errors not being displayed to users. When video transcoding failed (due to corrupted files, unsupported formats, or other issues), the error was only logged on the server with no feedback shown in the browser. Users would see indefinite loading spinners without knowing what went wrong. The system now properly communicates transcoding failures to your browser, displaying user-friendly error messages like "Video transcoding failed" so you know when a video can't be played. ([#244](https://github.com/djryanj/media-viewer/issues/244))
+
 ## [0.12.1] - 2026-02-10
 
 ### Fixed

--- a/internal/handlers/media_integration_test.go
+++ b/internal/handlers/media_integration_test.go
@@ -2311,3 +2311,278 @@ func TestListFilePathsPerformanceIntegration(t *testing.T) {
 		t.Errorf("expected %d items, got %d", numFiles, len(response.Items))
 	}
 }
+
+// =============================================================================
+// StreamVideo Error Handling Tests
+// =============================================================================
+
+// TestStreamVideo_TranscodingErrorIntegration tests that transcoding errors return HTTP 500
+func TestStreamVideo_TranscodingErrorIntegration(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping integration test")
+	}
+
+	// Create temporary directories
+	tempDir := t.TempDir()
+	dbPath := filepath.Join(tempDir, "test.db")
+	mediaDir := filepath.Join(tempDir, "media")
+	cacheDir := filepath.Join(tempDir, "cache")
+	thumbDir := filepath.Join(tempDir, "thumbnails")
+
+	if err := os.MkdirAll(mediaDir, 0o755); err != nil {
+		t.Fatalf("failed to create media dir: %v", err)
+	}
+	if err := os.MkdirAll(cacheDir, 0o755); err != nil {
+		t.Fatalf("failed to create cache dir: %v", err)
+	}
+	if err := os.MkdirAll(thumbDir, 0o755); err != nil {
+		t.Fatalf("failed to create thumb dir: %v", err)
+	}
+
+	// Create test video file that will need transcoding
+	testVideoPath := filepath.Join(mediaDir, "bad_video.flv")
+	if err := os.WriteFile(testVideoPath, []byte("fake corrupted video"), 0o644); err != nil {
+		t.Fatalf("Failed to create test video: %v", err)
+	}
+
+	// Mock ffprobe to return incompatible codec
+	mockFFProbe := filepath.Join(tempDir, "ffprobe")
+	ffprobeScript := `#!/bin/bash
+cat << 'EOF'
+{
+  "streams": [
+    {
+      "codec_name": "vp6f",
+      "width": 648,
+      "height": 459
+    }
+  ],
+  "format": {
+    "duration": "100.0"
+  }
+}
+EOF
+`
+	if err := os.WriteFile(mockFFProbe, []byte(ffprobeScript), 0o755); err != nil {
+		t.Fatalf("Failed to create mock ffprobe: %v", err)
+	}
+
+	// Mock ffmpeg to fail with realistic error
+	mockFFmpeg := filepath.Join(tempDir, "ffmpeg")
+	ffmpegScript := `#!/bin/bash
+cat << 'EOF' >&2
+ffmpeg version 8.0.1 Copyright (c) 2000-2025 the FFmpeg developers
+[libx264 @ 0x7f180d406040] height not divisible by 2 (648x459)
+[vost#0:0/libx264 @ 0x7f180a7e1c80] Error while opening encoder
+Conversion failed!
+EOF
+exit 187
+`
+	if err := os.WriteFile(mockFFmpeg, []byte(ffmpegScript), 0o755); err != nil {
+		t.Fatalf("Failed to create mock ffmpeg: %v", err)
+	}
+
+	// Temporarily modify PATH to use our mocks
+	oldPath := os.Getenv("PATH")
+	defer func() {
+		_ = os.Setenv("PATH", oldPath)
+	}()
+	_ = os.Setenv("PATH", tempDir+":"+oldPath)
+
+	// Create real database
+	db, err := database.New(context.Background(), dbPath)
+	if err != nil {
+		t.Fatalf("failed to create database: %v", err)
+	}
+	defer func() {
+		if err := db.Close(); err != nil {
+			t.Errorf("failed to close database: %v", err)
+		}
+	}()
+
+	// Create dependencies with transcoding ENABLED
+	idx := indexer.New(db, mediaDir, 0)
+	trans := transcoder.New(cacheDir, "", true)
+	thumbGen := media.NewThumbnailGenerator(cacheDir, mediaDir, false, db, 0, nil)
+
+	config := &startup.Config{
+		MediaDir: mediaDir,
+		CacheDir: cacheDir,
+	}
+
+	handlers := New(db, idx, trans, thumbGen, config)
+
+	// Create test request
+	req := httptest.NewRequest("GET", "/api/videos/bad_video.flv", http.NoBody)
+	w := httptest.NewRecorder()
+
+	// Add path variable
+	req = mux.SetURLVars(req, map[string]string{
+		"path": "bad_video.flv",
+	})
+
+	// Call StreamVideo handler
+	handlers.StreamVideo(w, req)
+
+	resp := w.Result()
+	defer resp.Body.Close()
+
+	// Verify HTTP 500 is returned
+	if resp.StatusCode != http.StatusInternalServerError {
+		t.Errorf("Expected status 500, got %d", resp.StatusCode)
+	}
+
+	// Verify error message in response
+	body := w.Body.String()
+	if !strings.Contains(body, "Failed to prepare video") {
+		t.Errorf("Expected error message about failed preparation, got: %s", body)
+	}
+
+	// Verify cache file was not created
+	cacheFiles, err := os.ReadDir(cacheDir)
+	if err != nil {
+		t.Fatalf("Failed to read cache dir: %v", err)
+	}
+
+	// Should only have error files, not actual cache files
+	for _, file := range cacheFiles {
+		if !strings.HasSuffix(file.Name(), ".err") && !strings.HasSuffix(file.Name(), ".tmp") {
+			t.Errorf("Unexpected cache file created: %s", file.Name())
+		}
+	}
+}
+
+// TestStreamVideo_CorruptedVideoErrorIntegration tests the exact error from issue logs
+func TestStreamVideo_CorruptedVideoErrorIntegration(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping integration test")
+	}
+
+	// This test simulates the exact error from the issue:
+	// [libx264 @ 0x7f180d406040] height not divisible by 2 (648x459)
+
+	tempDir := t.TempDir()
+	dbPath := filepath.Join(tempDir, "test.db")
+	mediaDir := filepath.Join(tempDir, "media")
+	cacheDir := filepath.Join(tempDir, "cache")
+
+	if err := os.MkdirAll(mediaDir, 0o755); err != nil {
+		t.Fatalf("failed to create media dir: %v", err)
+	}
+	if err := os.MkdirAll(cacheDir, 0o755); err != nil {
+		t.Fatalf("failed to create cache dir: %v", err)
+	}
+
+	// Create video file with odd dimensions that would trigger the error
+	testVideoPath := filepath.Join(mediaDir, "New Folder", "video.flv")
+	if err := os.MkdirAll(filepath.Dir(testVideoPath), 0o755); err != nil {
+		t.Fatalf("Failed to create directory: %v", err)
+	}
+	if err := os.WriteFile(testVideoPath, []byte("fake flv video"), 0o644); err != nil {
+		t.Fatalf("Failed to create test video: %v", err)
+	}
+
+	// Mock ffprobe to return the exact dimensions from the issue
+	mockFFProbe := filepath.Join(tempDir, "ffprobe")
+	ffprobeScript := `#!/bin/bash
+cat << 'EOF'
+{
+  "streams": [
+    {
+      "codec_name": "vp6f",
+      "width": 648,
+      "height": 459
+    }
+  ],
+  "format": {
+    "duration": "1385.82"
+  }
+}
+EOF
+`
+	if err := os.WriteFile(mockFFProbe, []byte(ffprobeScript), 0o755); err != nil {
+		t.Fatalf("Failed to create mock ffprobe: %v", err)
+	}
+
+	// Mock ffmpeg to fail with the exact error from the issue
+	mockFFmpeg := filepath.Join(tempDir, "ffmpeg")
+	ffmpegScript := `#!/bin/bash
+cat << 'EOF' >&2
+ffmpeg version 8.0.1 Copyright (c) 2000-2025 the FFmpeg developers
+built with gcc 15.2.0 (Alpine 15.2.0)
+Input #0, flv, from '/media/New Folder/video.flv':
+  Duration: 00:23:05.82, start: 0.000000, bitrate: 459 kb/s
+Stream #0:0: Audio: mp3 (mp3float), 22050 Hz, mono, fltp, 32 kb/s
+Stream #0:1: Video: vp6f, yuv420p, 648x459, 419 kb/s, 29.90 fps, 29.97 tbr, 1k tbn
+Stream mapping:
+  Stream #0:1 -> #0:0 (vp6f (native) -> h264 (libx264))
+  Stream #0:0 -> #0:1 (mp3 (mp3float) -> aac (native))
+Press [q] to stop, [?] for help
+[libx264 @ 0x7f180d406040] height not divisible by 2 (648x459)
+[vost#0:0/libx264 @ 0x7f180a7e1c80] Error while opening encoder - maybe incorrect parameters
+Error sending frames to consumers: Generic error in an external library
+Conversion failed!
+EOF
+exit 187
+`
+	if err := os.WriteFile(mockFFmpeg, []byte(ffmpegScript), 0o755); err != nil {
+		t.Fatalf("Failed to create mock ffmpeg: %v", err)
+	}
+
+	oldPath := os.Getenv("PATH")
+	defer func() {
+		_ = os.Setenv("PATH", oldPath)
+	}()
+	_ = os.Setenv("PATH", tempDir+":"+oldPath)
+
+	// Create dependencies
+	db, err := database.New(context.Background(), dbPath)
+	if err != nil {
+		t.Fatalf("failed to create database: %v", err)
+	}
+	defer func() {
+		if err := db.Close(); err != nil {
+			t.Errorf("failed to close database: %v", err)
+		}
+	}()
+
+	idx := indexer.New(db, mediaDir, 0)
+	trans := transcoder.New(cacheDir, "", true)
+	thumbGen := media.NewThumbnailGenerator(cacheDir, mediaDir, false, db, 0, nil)
+
+	config := &startup.Config{
+		MediaDir: mediaDir,
+		CacheDir: cacheDir,
+	}
+
+	handlers := New(db, idx, trans, thumbGen, config)
+
+	// Create request for the video
+	req := httptest.NewRequest("GET", "/api/videos/New%20Folder/video.flv", http.NoBody)
+	w := httptest.NewRecorder()
+
+	req = mux.SetURLVars(req, map[string]string{
+		"path": "New Folder/video.flv",
+	})
+
+	// Call handler
+	handlers.StreamVideo(w, req)
+
+	resp := w.Result()
+	defer resp.Body.Close()
+
+	// Verify HTTP 500 is returned (not hanging indefinitely)
+	if resp.StatusCode != http.StatusInternalServerError {
+		t.Errorf("Expected status 500 for incompatible video, got %d", resp.StatusCode)
+	}
+
+	// Verify response is timely (not waiting for 5 minute timeout)
+	// The test itself completing quickly proves this
+
+	body := w.Body.String()
+	t.Logf("Response body: %s", body)
+
+	if !strings.Contains(body, "Failed to prepare video") {
+		t.Errorf("Expected proper error message, got: %s", body)
+	}
+}

--- a/internal/transcoder/transcoder.go
+++ b/internal/transcoder/transcoder.go
@@ -157,6 +157,17 @@ func (t *Transcoder) GetVideoInfo(ctx context.Context, filePath string) (*VideoI
 		info.Height, _ = strconv.Atoi(heightStr)
 	}
 
+	// Tier 1: Ensure dimensions are even (required by H.264 encoder)
+	// Adjust odd dimensions to prevent encoding failures
+	if info.Width%2 != 0 {
+		logging.Debug("Adjusting odd width %d to %d for H.264 compatibility", info.Width, info.Width+1)
+		info.Width++
+	}
+	if info.Height%2 != 0 {
+		logging.Debug("Adjusting odd height %d to %d for H.264 compatibility", info.Height, info.Height+1)
+		info.Height++
+	}
+
 	// Check if transcoding is needed
 	ext := strings.ToLower(strings.TrimPrefix(filepath.Ext(filePath), "."))
 	info.NeedsTranscode = !compatibleCodecs[info.Codec] || !compatibleContainers[ext]
@@ -198,6 +209,10 @@ func (t *Transcoder) GetOrStartTranscode(_ context.Context, filePath string, tar
 		return cachePath, true, nil
 	}
 
+	// Clean up any stale error file from previous failed attempts
+	errorPath := cachePath + ".err"
+	_ = os.Remove(errorPath)
+
 	// Start background transcoding
 	logging.Info("Starting background transcode: %s -> %s", filePath, cachePath)
 
@@ -213,11 +228,18 @@ func (t *Transcoder) GetOrStartTranscode(_ context.Context, filePath string, tar
 
 		if err := t.transcodeDirectToCache(bgCtx, filePath, cachePath, targetWidth, info, needsReencode); err != nil {
 			logging.Error("Background transcode failed for %s: %v", filePath, err)
+			// Write error to .err file so other code can detect it
+			errorPath := cachePath + ".err"
+			if writeErr := os.WriteFile(errorPath, []byte(err.Error()), 0o600); writeErr != nil {
+				logging.Warn("Failed to write error file: %v", writeErr)
+			}
 			// Clean up partial cache on error
 			_ = os.Remove(cachePath)
 			_ = os.Remove(cachePath + ".tmp")
 		} else {
 			logging.Info("Background transcode completed: %s", cachePath)
+			// Remove any error file from previous failed attempts
+			_ = os.Remove(cachePath + ".err")
 		}
 	}()
 
@@ -257,6 +279,10 @@ func (t *Transcoder) GetOrStartTranscodeAndWait(ctx context.Context, filePath st
 			return cachePath, nil
 		}
 
+		// Clean up any stale error file from previous failed attempts
+		errorPath := cachePath + ".err"
+		_ = os.Remove(errorPath)
+
 		// Start background transcoding
 		logging.Info("Starting background transcode: %s -> %s", filePath, cachePath)
 
@@ -272,11 +298,18 @@ func (t *Transcoder) GetOrStartTranscodeAndWait(ctx context.Context, filePath st
 
 			if err := t.transcodeDirectToCache(bgCtx, filePath, cachePath, targetWidth, info, needsReencode); err != nil {
 				logging.Error("Background transcode failed for %s: %v", filePath, err)
+				// Write error to .err file so waiting code can detect it
+				errorPath := cachePath + ".err"
+				if writeErr := os.WriteFile(errorPath, []byte(err.Error()), 0o600); writeErr != nil {
+					logging.Warn("Failed to write error file: %v", writeErr)
+				}
 				// Clean up partial cache on error
 				_ = os.Remove(cachePath)
 				_ = os.Remove(cachePath + ".tmp")
 			} else {
 				logging.Info("Background transcode completed: %s", cachePath)
+				// Remove any error file from previous failed attempts
+				_ = os.Remove(cachePath + ".err")
 			}
 		}()
 	} else {
@@ -294,6 +327,14 @@ func (t *Transcoder) GetOrStartTranscodeAndWait(ctx context.Context, filePath st
 
 	for {
 		tmpPath := cachePath + ".tmp"
+		errorPath := cachePath + ".err"
+
+		// Check if transcoding failed (error file exists)
+		if errorData, err := os.ReadFile(errorPath); err == nil {
+			// Clean up error file
+			_ = os.Remove(errorPath)
+			return "", fmt.Errorf("transcode failed: %s", string(errorData))
+		}
 
 		// Check if transcode is complete (.tmp file gone)
 		if _, err := os.Stat(tmpPath); os.IsNotExist(err) {
@@ -301,6 +342,8 @@ func (t *Transcoder) GetOrStartTranscodeAndWait(ctx context.Context, filePath st
 			if stat, err := os.Stat(cachePath); err == nil {
 				logging.Info("Transcode complete: %.2f MB for %s (took %.1fs)",
 					float64(stat.Size())/(1024*1024), filePath, time.Since(startWait).Seconds())
+				// Clean up any stale error files
+				_ = os.Remove(errorPath)
 				return cachePath, nil
 			}
 		} else {
@@ -835,10 +878,17 @@ func (t *Transcoder) buildFFmpegArgs(inputPath, outputPath string, targetWidth i
 		// Re-encode with h264
 		args = append(args, "-c:v", "libx264", "-preset", "fast", "-crf", "23")
 
-		// Add scale filter if needed
+		// Tier 2: Always add scale filter when re-encoding to ensure output dimensions
+		// match the (possibly adjusted) dimensions from GetVideoInfo
 		if needsScaling {
+			// Scale to requested width, maintaining aspect ratio with even height
 			logging.Debug("Adding scale filter: %dx-2", targetWidth)
 			args = append(args, "-vf", fmt.Sprintf("scale=%d:-2", targetWidth))
+		} else {
+			// No size reduction, but force exact dimensions to handle odd dimensions
+			// This ensures output matches Tier 1 adjusted dimensions (always even)
+			logging.Debug("Adding scale filter for exact dimensions: %dx%d", info.Width, info.Height)
+			args = append(args, "-vf", fmt.Sprintf("scale=%d:%d", info.Width, info.Height))
 		}
 	}
 

--- a/internal/transcoder/transcoder_coverage_test.go
+++ b/internal/transcoder/transcoder_coverage_test.go
@@ -203,6 +203,114 @@ sleep 10
 	}
 }
 
+func TestGetVideoInfo_AdjustsOddWidth(t *testing.T) {
+	tmpDir := t.TempDir()
+	mockFFProbe := filepath.Join(tmpDir, "ffprobe")
+
+	// Return video with odd width (1919)
+	ffprobeScript := `#!/bin/bash
+echo '{"streams":[{"codec_name":"h264","width":1919,"height":1080}],"format":{"duration":"100.0"}}'
+`
+
+	if err := os.WriteFile(mockFFProbe, []byte(ffprobeScript), 0o755); err != nil {
+		t.Fatalf("Failed to create mock ffprobe: %v", err)
+	}
+
+	oldPath := os.Getenv("PATH")
+	defer func() {
+		_ = os.Setenv("PATH", oldPath)
+	}()
+	_ = os.Setenv("PATH", tmpDir+":"+oldPath)
+
+	trans := New("/tmp/cache", "", true)
+	ctx := context.Background()
+
+	info, err := trans.GetVideoInfo(ctx, "/fake/video.mp4")
+	if err != nil {
+		t.Fatalf("GetVideoInfo() error: %v", err)
+	}
+
+	// Width should be adjusted from 1919 to 1920
+	if info.Width != 1920 {
+		t.Errorf("Expected width=1920 (adjusted from 1919), got %d", info.Width)
+	}
+	if info.Height != 1080 {
+		t.Errorf("Expected height=1080 (unchanged), got %d", info.Height)
+	}
+}
+
+func TestGetVideoInfo_AdjustsOddHeight(t *testing.T) {
+	tmpDir := t.TempDir()
+	mockFFProbe := filepath.Join(tmpDir, "ffprobe")
+
+	// Return video with odd height (459) - the actual failure case from the log
+	ffprobeScript := `#!/bin/bash
+echo '{"streams":[{"codec_name":"vp6f","width":648,"height":459}],"format":{"duration":"100.0"}}'
+`
+
+	if err := os.WriteFile(mockFFProbe, []byte(ffprobeScript), 0o755); err != nil {
+		t.Fatalf("Failed to create mock ffprobe: %v", err)
+	}
+
+	oldPath := os.Getenv("PATH")
+	defer func() {
+		_ = os.Setenv("PATH", oldPath)
+	}()
+	_ = os.Setenv("PATH", tmpDir+":"+oldPath)
+
+	trans := New("/tmp/cache", "", true)
+	ctx := context.Background()
+
+	info, err := trans.GetVideoInfo(ctx, "/fake/video.flv")
+	if err != nil {
+		t.Fatalf("GetVideoInfo() error: %v", err)
+	}
+
+	// Height should be adjusted from 459 to 460
+	if info.Width != 648 {
+		t.Errorf("Expected width=648 (unchanged), got %d", info.Width)
+	}
+	if info.Height != 460 {
+		t.Errorf("Expected height=460 (adjusted from 459), got %d", info.Height)
+	}
+}
+
+func TestGetVideoInfo_AdjustsBothOddDimensions(t *testing.T) {
+	tmpDir := t.TempDir()
+	mockFFProbe := filepath.Join(tmpDir, "ffprobe")
+
+	// Return video with both odd dimensions
+	ffprobeScript := `#!/bin/bash
+echo '{"streams":[{"codec_name":"h264","width":1279,"height":719}],"format":{"duration":"100.0"}}'
+`
+
+	if err := os.WriteFile(mockFFProbe, []byte(ffprobeScript), 0o755); err != nil {
+		t.Fatalf("Failed to create mock ffprobe: %v", err)
+	}
+
+	oldPath := os.Getenv("PATH")
+	defer func() {
+		_ = os.Setenv("PATH", oldPath)
+	}()
+	_ = os.Setenv("PATH", tmpDir+":"+oldPath)
+
+	trans := New("/tmp/cache", "", true)
+	ctx := context.Background()
+
+	info, err := trans.GetVideoInfo(ctx, "/fake/video.mp4")
+	if err != nil {
+		t.Fatalf("GetVideoInfo() error: %v", err)
+	}
+
+	// Both dimensions should be adjusted
+	if info.Width != 1280 {
+		t.Errorf("Expected width=1280 (adjusted from 1279), got %d", info.Width)
+	}
+	if info.Height != 720 {
+		t.Errorf("Expected height=720 (adjusted from 719), got %d", info.Height)
+	}
+}
+
 // =============================================================================
 // StreamVideo Tests
 // =============================================================================
@@ -980,6 +1088,123 @@ func TestTranscodeAndStreamIntegration_FullProcess(t *testing.T) {
 	t.Logf("Transcoded %d bytes at %d width", buf.Len(), targetWidth)
 }
 
+func TestTranscodeIntegration_OddDimensions(t *testing.T) {
+	if testing.Short() {
+		t.Skip("Skipping integration test in short mode")
+	}
+
+	tmpDir := t.TempDir()
+	cacheDir := filepath.Join(tmpDir, "cache")
+	if err := os.MkdirAll(cacheDir, 0o755); err != nil {
+		t.Fatalf("Failed to create cache dir: %v", err)
+	}
+
+	// Create a test video with odd dimensions (648x459) - the actual failure case
+	// Use mpeg4 codec which allows odd dimensions during creation
+	testFile := filepath.Join(tmpDir, "odd_dimensions.mp4")
+	cmd := exec.CommandContext(context.Background(), "ffmpeg",
+		"-f", "lavfi",
+		"-i", "testsrc=duration=1:size=648x459:rate=1",
+		"-c:v", "mpeg4", // mpeg4 allows odd dimensions
+		"-t", "1",
+		testFile,
+	)
+
+	if err := cmd.Run(); err != nil {
+		t.Skipf("Skipping test: ffmpeg not available or failed: %v", err)
+	}
+
+	trans := New(cacheDir, "", true)
+	ctx := context.Background()
+
+	// Get video info - should adjust odd dimensions
+	info, err := trans.GetVideoInfo(ctx, testFile)
+	if err != nil {
+		t.Fatalf("GetVideoInfo failed: %v", err)
+	}
+
+	// Verify dimensions were adjusted
+	if info.Height%2 != 0 {
+		t.Errorf("Height should be adjusted to even number, got %d", info.Height)
+	}
+	if info.Width%2 != 0 {
+		t.Errorf("Width should be adjusted to even number, got %d", info.Width)
+	}
+
+	// Now transcode it - this should succeed without "height not divisible by 2" error
+	var buf bytes.Buffer
+	err = trans.StreamVideo(ctx, testFile, &buf, 0) // targetWidth=0 means no scaling
+	if err != nil {
+		t.Fatalf("StreamVideo failed with odd dimensions: %v", err)
+	}
+
+	if buf.Len() == 0 {
+		t.Error("Expected transcoded video data")
+	}
+
+	t.Logf("Successfully transcoded video with originally odd dimensions, output: %d bytes", buf.Len())
+}
+
+func TestTranscodeIntegration_OddDimensionsIncompatibleCodec(t *testing.T) {
+	if testing.Short() {
+		t.Skip("Skipping integration test in short mode")
+	}
+
+	tmpDir := t.TempDir()
+	cacheDir := filepath.Join(tmpDir, "cache")
+	if err := os.MkdirAll(cacheDir, 0o755); err != nil {
+		t.Fatalf("Failed to create cache dir: %v", err)
+	}
+
+	// Create a test video with odd dimensions using an incompatible codec (HEVC)
+	// This tests the case where needsReencode=true but needsScaling=false
+	testFile := filepath.Join(tmpDir, "odd_hevc.mp4")
+	cmd := exec.CommandContext(context.Background(), "ffmpeg",
+		"-f", "lavfi",
+		"-i", "testsrc=duration=1:size=320x239:rate=1", // 239 is odd
+		"-c:v", "libx265", // HEVC/H.265 codec (incompatible for web)
+		"-t", "1",
+		testFile,
+	)
+
+	if err := cmd.Run(); err != nil {
+		t.Skipf("Skipping test: ffmpeg/libx265 not available or failed: %v", err)
+	}
+
+	trans := New(cacheDir, "", true)
+	ctx := context.Background()
+
+	// Get video info
+	info, err := trans.GetVideoInfo(ctx, testFile)
+	if err != nil {
+		t.Fatalf("GetVideoInfo failed: %v", err)
+	}
+
+	// Verify height was adjusted to even
+	if info.Height%2 != 0 {
+		t.Errorf("Height should be adjusted to even number, got %d", info.Height)
+	}
+
+	// HEVC should need transcode (incompatible codec forces re-encode)
+	if !info.NeedsTranscode {
+		t.Error("HEVC codec should need transcode for H.264 compatibility")
+	}
+
+	// Transcode without scaling (targetWidth=0)
+	// This triggers Tier 2: re-encode with scale filter to ensure even dimensions
+	var buf bytes.Buffer
+	err = trans.StreamVideo(ctx, testFile, &buf, 0)
+	if err != nil {
+		t.Fatalf("StreamVideo failed for incompatible codec with odd dimensions: %v", err)
+	}
+
+	if buf.Len() == 0 {
+		t.Error("Expected transcoded video data")
+	}
+
+	t.Logf("Successfully re-encoded HEVC video with odd dimensions to H.264, output: %d bytes", buf.Len())
+}
+
 func TestCleanupIntegration_WithActiveTranscode(t *testing.T) {
 	checkFFmpegAvailable(t)
 	testVideo := getTestVideoPath(t)
@@ -1156,4 +1381,261 @@ func createTestH264Video(t *testing.T, dir string) string {
 	}
 
 	return videoPath
+}
+
+// =============================================================================
+// Error Handling Tests
+// =============================================================================
+
+func TestGetOrStartTranscodeAndWait_ErrorFileDetection(t *testing.T) {
+	tmpDir := t.TempDir()
+	cacheDir := filepath.Join(tmpDir, "cache")
+	if err := os.MkdirAll(cacheDir, 0o755); err != nil {
+		t.Fatalf("Failed to create cache dir: %v", err)
+	}
+
+	// Create test video file
+	testFile := filepath.Join(tmpDir, "test.mkv")
+	if err := os.WriteFile(testFile, []byte("fake video"), 0o644); err != nil {
+		t.Fatalf("Failed to create test file: %v", err)
+	}
+
+	// Mock ffprobe to return incompatible codec
+	mockFFProbe := filepath.Join(tmpDir, "ffprobe")
+	ffprobeScript := `#!/bin/bash
+cat << 'EOF'
+{
+  "streams": [
+    {
+      "codec_name": "hevc",
+      "width": 1920,
+      "height": 1080
+    }
+  ],
+  "format": {
+    "duration": "100.0"
+  }
+}
+EOF
+`
+	if err := os.WriteFile(mockFFProbe, []byte(ffprobeScript), 0o755); err != nil {
+		t.Fatalf("Failed to create mock ffprobe: %v", err)
+	}
+
+	// Mock ffmpeg to fail
+	mockFFmpeg := filepath.Join(tmpDir, "ffmpeg")
+	ffmpegScript := `#!/bin/bash
+echo "ffmpeg version 8.0 mock error" >&2
+echo "[libx264 @ 0x123] height not divisible by 2 (1920x1079)" >&2
+echo "Conversion failed!" >&2
+exit 1
+`
+	if err := os.WriteFile(mockFFmpeg, []byte(ffmpegScript), 0o755); err != nil {
+		t.Fatalf("Failed to create mock ffmpeg: %v", err)
+	}
+
+	oldPath := os.Getenv("PATH")
+	defer func() {
+		_ = os.Setenv("PATH", oldPath)
+	}()
+	_ = os.Setenv("PATH", tmpDir+":"+oldPath)
+
+	trans := New(cacheDir, "", true)
+	ctx := context.Background()
+
+	// Get video info
+	info, err := trans.GetVideoInfo(ctx, testFile)
+	if err != nil {
+		t.Fatalf("GetVideoInfo failed: %v", err)
+	}
+
+	// This should start transcoding and wait, but ffmpeg will fail
+	// The error should be detected via the .err file
+	cachePath, err := trans.GetOrStartTranscodeAndWait(ctx, testFile, 0, info)
+
+	// Verify error is returned
+	if err == nil {
+		t.Error("Expected error from failed transcode, got nil")
+	}
+
+	if cachePath != "" {
+		t.Errorf("Expected empty cachePath on error, got: %s", cachePath)
+	}
+
+	// Verify error message contains transcoding failure info
+	if !strings.Contains(err.Error(), "transcode failed") {
+		t.Errorf("Expected 'transcode failed' in error, got: %v", err)
+	}
+
+	// Verify cache file doesn't exist
+	cacheKey := filepath.Base(testFile) + "_w0.mp4"
+	finalCachePath := filepath.Join(cacheDir, cacheKey)
+	if _, err := os.Stat(finalCachePath); !os.IsNotExist(err) {
+		t.Error("Cache file should not exist after failed transcode")
+	}
+
+	// Verify error file is cleaned up
+	errorPath := finalCachePath + ".err"
+	if _, err := os.Stat(errorPath); !os.IsNotExist(err) {
+		t.Error("Error file should be cleaned up after detection")
+	}
+}
+
+func TestGetOrStartTranscode_ErrorFileCreation(t *testing.T) {
+	tmpDir := t.TempDir()
+	cacheDir := filepath.Join(tmpDir, "cache")
+	if err := os.MkdirAll(cacheDir, 0o755); err != nil {
+		t.Fatalf("Failed to create cache dir: %v", err)
+	}
+
+	// Create test video file
+	testFile := filepath.Join(tmpDir, "test.mkv")
+	if err := os.WriteFile(testFile, []byte("fake video"), 0o644); err != nil {
+		t.Fatalf("Failed to create test file: %v", err)
+	}
+
+	// Mock ffprobe to return incompatible codec
+	mockFFProbe := filepath.Join(tmpDir, "ffprobe")
+	ffprobeScript := `#!/bin/bash
+cat << 'EOF'
+{
+  "streams": [
+    {
+      "codec_name": "hevc",
+      "width": 1920,
+      "height": 1080
+    }
+  ],
+  "format": {
+    "duration": "100.0"
+  }
+}
+EOF
+`
+	if err := os.WriteFile(mockFFProbe, []byte(ffprobeScript), 0o755); err != nil {
+		t.Fatalf("Failed to create mock ffprobe: %v", err)
+	}
+
+	// Mock ffmpeg to fail
+	mockFFmpeg := filepath.Join(tmpDir, "ffmpeg")
+	ffmpegScript := `#!/bin/bash
+echo "Fatal error in transcoding" >&2
+exit 187
+`
+	if err := os.WriteFile(mockFFmpeg, []byte(ffmpegScript), 0o755); err != nil {
+		t.Fatalf("Failed to create mock ffmpeg: %v", err)
+	}
+
+	oldPath := os.Getenv("PATH")
+	defer func() {
+		_ = os.Setenv("PATH", oldPath)
+	}()
+	_ = os.Setenv("PATH", tmpDir+":"+oldPath)
+
+	trans := New(cacheDir, "", true)
+	ctx := context.Background()
+
+	// Get video info
+	info, err := trans.GetVideoInfo(ctx, testFile)
+	if err != nil {
+		t.Fatalf("GetVideoInfo failed: %v", err)
+	}
+
+	// Start transcoding (won't wait)
+	cachePath, isCached, err := trans.GetOrStartTranscode(ctx, testFile, 0, info)
+	if err != nil {
+		t.Fatalf("GetOrStartTranscode failed: %v", err)
+	}
+
+	if isCached {
+		t.Error("Expected isCached=false for new transcode")
+	}
+
+	// Wait for the background goroutine to finish and create error file
+	time.Sleep(500 * time.Millisecond)
+
+	// Verify error file was created
+	errorPath := cachePath + ".err"
+	errorData, err := os.ReadFile(errorPath)
+	if err != nil {
+		t.Errorf("Error file should exist at %s: %v", errorPath, err)
+	} else {
+		errorStr := string(errorData)
+		if !strings.Contains(errorStr, "ffmpeg error") {
+			t.Errorf("Error file should contain ffmpeg error, got: %s", errorStr)
+		}
+		if !strings.Contains(errorStr, "187") {
+			t.Errorf("Error file should contain exit code 187, got: %s", errorStr)
+		}
+	}
+
+	// Verify cache files were cleaned up
+	if _, err := os.Stat(cachePath); !os.IsNotExist(err) {
+		t.Error("Cache file should be removed on error")
+	}
+
+	tmpPath := cachePath + ".tmp"
+	if _, err := os.Stat(tmpPath); !os.IsNotExist(err) {
+		t.Error("Temp cache file should be removed on error")
+	}
+}
+
+func TestGetOrStartTranscodeAndWait_SuccessfulTranscodeCleansErrorFile(t *testing.T) {
+	if testing.Short() {
+		t.Skip("Skipping integration test in short mode")
+	}
+
+	tmpDir := t.TempDir()
+	cacheDir := filepath.Join(tmpDir, "cache")
+	if err := os.MkdirAll(cacheDir, 0o755); err != nil {
+		t.Fatalf("Failed to create cache dir: %v", err)
+	}
+
+	// Create a small test video using ffmpeg (if available)
+	testFile := filepath.Join(tmpDir, "test.mp4")
+	cmd := exec.CommandContext(context.Background(), "ffmpeg",
+		"-f", "lavfi",
+		"-i", "testsrc=duration=1:size=320x240:rate=1",
+		"-c:v", "libx264",
+		"-t", "1",
+		"-pix_fmt", "yuv420p",
+		testFile,
+	)
+
+	if err := cmd.Run(); err != nil {
+		t.Skipf("Skipping test: ffmpeg not available or failed: %v", err)
+	}
+
+	trans := New(cacheDir, "", true)
+	ctx := context.Background()
+
+	// Get video info
+	info, err := trans.GetVideoInfo(ctx, testFile)
+	if err != nil {
+		t.Fatalf("GetVideoInfo failed: %v", err)
+	}
+
+	// Create a fake error file from a previous failed attempt
+	cacheKey := filepath.Base(testFile) + "_w0.mp4"
+	cachePath := filepath.Join(cacheDir, cacheKey)
+	errorPath := cachePath + ".err"
+	if err := os.WriteFile(errorPath, []byte("previous error"), 0o644); err != nil {
+		t.Fatalf("Failed to create fake error file: %v", err)
+	}
+
+	// Verify error file exists
+	if _, err := os.Stat(errorPath); err != nil {
+		t.Fatal("Error file should exist before test")
+	}
+
+	// This should succeed and clean up the error file
+	_, err = trans.GetOrStartTranscodeAndWait(ctx, testFile, 0, info)
+	if err != nil {
+		t.Fatalf("GetOrStartTranscodeAndWait failed: %v", err)
+	}
+
+	// Verify error file was removed
+	if _, err := os.Stat(errorPath); !os.IsNotExist(err) {
+		t.Error("Error file should be removed after successful transcode")
+	}
 }

--- a/static/js/lightbox.js
+++ b/static/js/lightbox.js
@@ -1739,7 +1739,7 @@ const Lightbox = {
             console.error('Error loading video:', e);
             this.hideLoading();
 
-            // Check if this is an auth error
+            // Check if this is an auth error or other server error
             try {
                 const response = await fetchWithTimeout(videoUrl, {
                     method: 'HEAD',
@@ -1751,6 +1751,24 @@ const Lightbox = {
                         SessionManager.handleSessionExpired();
                     } else {
                         window.location.replace('/login.html');
+                    }
+                } else if (response.status === 500) {
+                    console.error('Lightbox: server error loading video');
+                    if (typeof Gallery !== 'undefined' && Gallery.showToast) {
+                        Gallery.showToast(
+                            'Failed to load video. The file may be corrupted or incompatible with transcoding.',
+                            'error',
+                            10000
+                        );
+                    }
+                } else if (response.status >= 400) {
+                    console.error('Lightbox: HTTP error', response.status);
+                    if (typeof Gallery !== 'undefined' && Gallery.showToast) {
+                        Gallery.showToast(
+                            `Failed to load video (Error ${response.status})`,
+                            'error',
+                            8000
+                        );
                     }
                 }
             } catch (err) {

--- a/static/js/playlist.js
+++ b/static/js/playlist.js
@@ -1056,7 +1056,7 @@ const Playlist = {
     },
 
     /**
-     * Check if a video load failure was due to authentication
+     * Check if a video load failure was due to authentication or other server errors
      * @param {string} videoUrl - The URL that failed to load
      */
     async checkVideoAuthError(videoUrl) {
@@ -1069,6 +1069,24 @@ const Playlist = {
                 } else {
                     this.close();
                     window.location.replace('/login.html');
+                }
+            } else if (response.status === 500) {
+                console.error('Player: server error loading video');
+                if (typeof Gallery !== 'undefined' && Gallery.showToast) {
+                    Gallery.showToast(
+                        'Failed to load video. The file may be corrupted or incompatible with transcoding.',
+                        'error',
+                        10000
+                    );
+                }
+            } else if (response.status >= 400) {
+                console.error('Player: HTTP error', response.status);
+                if (typeof Gallery !== 'undefined' && Gallery.showToast) {
+                    Gallery.showToast(
+                        `Failed to load video (Error ${response.status})`,
+                        'error',
+                        8000
+                    );
                 }
             }
         } catch (e) {

